src/how-to-build-and-run.md

```diff
@@ -140,6 +160,63 @@ release: 1.25.0-dev
 LLVM version: 4.0
 ```
 
+<a name=workflow>
+
+### Suggested workflows for faster builds of the compiler
+
+There are two workflows that are useful for faster builders of the
+compiler.
+
+**Check, check, and check again.** The first workflow, which is useful
+when doing simple refactorings, is to run `./x.py check`
+continuously. Here you are just checking that the compiler can
+**build**, but often that is all you need (e.g., when renaming a
+method). You can then run `./x.py build` when you actually need to
+run tests.
+
+In fact, it is sometimes useful to put off tests even when you are not
+100% sure the code will work. You can then keep building up
+refactoring commits and only run the tests at some later time. You can
+then use `git bisect` to track down **precisely** which commit caused
+the problem. A nice side-effect of this style is that you are left
+with a fairly fine-grained set of commits at the end, all of which
+build and pass testes. This often helps reviewing.
+
+**Incremental builds with `--keep-stage`.** Sometimes just checking
+whether the compiler builds is not enough. A common example is that
+you need to add a `debug!` statement to inspect the value of some
+state or better understand the problem. In that case, you really need
+a full build.  By leveraging incremental, though, you can often get
+these builds to complete very fast (e.g., around 30 seconds): the only
+catch is this requires a bit of fudging and may produce compilers that
+don't work (but that is easily detected and fixed).
+
+The sequence of commands you want is as follows:
+
+- Initial build: `./x.py build -i --stage 1 src/libstd`
+  - As [documented above](#command), this will build a functional
+    stage1 compiler
+- Subsequent builds: `./x.py build -i --stage 1 src/libstd --keep-stage 1`
+  - Note that we added the `--keep-stage 1` flag here
+  
+The effect of `--keep-stage1` is that we just *assume* that the old
+standard library can be re-used. If you are editing the compiler, this
+is almost always true: you haven't changed the standard library, after
+all.  But sometimes, it's not true: for example, if you are editing
+the "metadata" part of the compiler, which controls how the compiler
+encodes types and other states into the `rlib` files, or if you are editing
+things that wind up in the metadata (such as the definition of the MIR).
+
+**The TL;DR is that you might get weird behavior from a compile when
+using `--keep-stage 1`** -- for example, strange ICEs or other
+panics. In that case, you should simply remove the `--keep-stage 1`
+from the command and rebuild.  That ought to fix the problem.
+
+Note: you can also use `--keep-stage 1` when running tests. Something like this:
+
+- Initial test run: `./x.py test -i --stage 1 src/test/ui`
+- Subsequent test run: `./x.py test -i --stage 1 src/test/ui --keep-stage 1`
+
```

Comment:
I think it's good to note that the `--keep-stage` flag can be passed multiple times.

Furthermore, if you want guaranteed to work and optimal build times when changing just the compiler, the workflow I would suggest is: `./x.py build -i --stage 2 --keep-stage 0`; the `--stage 2` is optional. The effect of this command is that when the compiler is changed, we just rebuild it and you have a compiler that is now both fully capable (i.e., proc macros, sysroot fully populated) and you did not need to rebuild std at all. If, however, you wanted to have your changes take effect on the std that compiler uses, then that won't happen with this. But that's relatively rare.

Reply:
> and you did not need to rebuild std at all

I don't think such a compiler can possibly be guaranteed to work. I guess what you mean is "if you want a compiler that avoids the stage1 limitations"?

Reply:
I decided not to add this text because I still don't really understand what `--keep-stage 0` even means. :)  Does it mean that we copy the stdlib from the beta?

I guess in that case maybe we *always* want to suggest it, unless you are editing the stdlib?

Reply:
cc @Mark-Simulacrum